docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,334 @@
+  the cache provided to store the bear results. The default value of this
+  parameter is ``None`` which when provided, runs coala without a cache.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+
+Bears in the NextGen-Core are implemented differently as compared to the old
+bears. For instance every bear has an ``analyze`` function to perform code
+analysis instead of the ``run`` function that was there in the old bears.
+Following points must be kept in mind while writing NextGen bears.
+
+* The new bears always have to constructed with ``section`` and ``file_dict``
+  as parameters, this is important as it makes the new dependency system
+  possible.
+* Bear writers need to keep this in mind while writing their ``__init__``
+  methods, default parameters are allowed as many as wanted but the same isn't
+  true for positional parameters (default parameters are discouraged, since the
+  dependency system automatically instantiates bears, which might lead to
+  unwanted side-effects). More details can be found at the
+  `API Docs <http://api.coala.io/>`_.
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files. Bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs.
+
+Task Objects
+------------
+
+Task objects are the representation of tasks performed by bears. Structure-wise
+they are a tuple containing the tuples of positional arguments and a dicts of
+keyword arguments to the ``analyze`` (which are being invoked by the
+``execute_task`` function) function. Task objects are instrumental to the
+caching mechanism as their hash values are stored in the cache along with the
+bear results and are looked up during each coala run to fetch the results.
+
+For instance we have a ``FileBear`` that looks like this:
+
+::
+
+    from coalib.core.FileBear import FileBear
+
+
+    class SomeFileBear(FileBear):
+
+        def analyze(self, file, filename, filename_prefix: str='',
+                    filename_suffix: str=''):
+            yield 'Some analysis result'
+
+Its corresponding task object would look like the one below:
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+    ]
+
+Improved Dependency System
+--------------------------
+
+The NextGen-Core introduces a better dependency management system than the
+one used by the old core. It features following improvements:
+
+*  A bear dependent on other bears can specify its dependencies in
+   ``BEAR_DEPS``.
```

Comment:
I agree with you that an implementation detail like this must not be mentioned as a feature but since it resides in the `Improved Dependency System` section which itself is a featured improvement I think we can have this little implementation detail as a part of it. I think it should stay(we are not saying that its new its just part of something new).